crates/spirv-std/src/image.rs

```diff
@@ -659,6 +659,185 @@ impl<
+    #[crate::macros::gpu_only]
+    #[doc(alias = "OpImageQueryLevels")]
+    pub fn query_levels(&self) -> u32 {
+        let result: u32;
+        unsafe {
+            asm! {
+                "%image = OpLoad _ {this}",
+                "{result} = OpImageQueryLevels typeof{result} %image",
+                this = in(reg) self,
+                result = out(reg) result,
+            }
+        }
+        result
+    }
+
+    /// Query the mipmap level and the level of detail for a hypothetical sampling of Image at
+    /// Coordinate using an implicit level of detail. The first component of the result contains
+    /// the mipmap array layer. The second component of the result contains the implicit level of
+    /// detail relative to the base level.
+    ///
+    /// Note: Const generics aren't able to reason about the constraints on this function (yet),
+    /// and so are enforced by the compiler. The constraints are:
+    ///
+    /// The image's dimension must be 1D, 2D, 3D, or Cube.
+    #[crate::macros::gpu_only]
+    #[doc(alias = "OpImageQueryLod")]
+    pub fn query_lod<V: Vector<f32, 2>>(
+        &self,
+        sampler: Sampler,
+        coord: impl ImageCoordinate<f32, DIM, { Arrayed::False }>,
+    ) -> V {
+        // Note: Arrayed::False isn't a typo in the ImageCoordinate, the spec states:
+        // Coordinate must be a scalar or vector of floating-point type or integer type. It
+        // contains (u[, v] ... ) as needed by the definition of Sampled Image, **not including any
+        // array layer index**. Unless the Kernel capability is being used, it must be floating
+        // point.
+        let mut result = Default::default();
+        unsafe {
+            asm! {
+                "%typeSampledImage = OpTypeSampledImage typeof*{this}",
+                "%image = OpLoad _ {this}",
+                "%sampler = OpLoad _ {sampler}",
+                "%coord = OpLoad _ {coord}",
+                "%sampledImage = OpSampledImage %typeSampledImage %image %sampler",
+                "%result = OpImageQueryLod typeof*{result} %sampledImage %coord",
+                "OpStore {result} %result",
+                result = in(reg) &mut result,
+                this = in(reg) self,
+                sampler = in(reg) &sampler,
+                coord = in(reg) &coord
+            }
+        }
+        result
+    }
+
+    /// Query the dimensions of Image, with no level of detail.
+    ///
+    /// Note: Const generics aren't able to reason about the constraints on this function (yet),
+    /// and so are enforced by the compiler. The constraints are:
+    ///
+    /// The image's dimension must be 1D, 2D, 3D, Buffer, Rect, or Cube. If dimension is 1D, 2D,
+    /// 3D, or Cube, it must have *either* multisampled be true, *or* sampled of Unknown or No.
+    #[crate::macros::gpu_only]
+    #[doc(alias = "OpImageQuerySize")]
+    pub fn query_size<Size: ImageCoordinate<u32, DIM, ARRAYED> + Default>(&self) -> Size {
```

Comment:
`ImageCoordinate` itself could probably require all implementations to also have Default as the only valid image coordinate types are vectors and scalars all of which impl `Default`.